## 异常介绍

异常就是程序运行时发生错误的信号，在python中,错误触发的异常如下：

![](http://omk1n04i8.bkt.clouddn.com/17-8-1/22612195.jpg)

### 异常种类

在python中不同的异常可以用不同的类型（python中统一了类与类型，类型即类）去标识，不同的类对象标识不同的异常，一个异常标识一种错误：

```
AttributeError 试图访问一个对象没有的属性，比如foo.x，但是foo没有属性x
IOError 输入/输出异常；基本上是无法打开文件
ImportError 无法引入模块或包；基本上是路径问题或名称错误
IndentationError 语法错误（的子类） ；代码没有正确对齐，缩进错误
IndexError 下标索引超出序列边界，比如当x只有三个元素，却试图访问x[5]
KeyError 试图访问字典里不存在的键
KeyboardInterrupt Ctrl+C被按下
NameError 使用一个还未被赋予对象的变量
SyntaxError Python代码非法，代码不能编译(个人认为这是语法错误，写错了）
TypeError 传入对象类型与要求的不符合
UnboundLocalError 试图访问一个还未被设置的局部变量，基本上是由于另有一个同名的全局变量，
导致你以为正在访问它
ValueError 传入一个调用者不期望的值，即使值的类型是正确的
```

更多异常：

```
ArithmeticError
AssertionError
AttributeError
BaseException
BufferError
BytesWarning
DeprecationWarning
EnvironmentError
EOFError
Exception
FloatingPointError
FutureWarning
GeneratorExit
ImportError
ImportWarning
IndentationError
IndexError
IOError
KeyboardInterrupt
KeyError
LookupError
MemoryError
NameError
NotImplementedError
OSError
OverflowError
PendingDeprecationWarning
ReferenceError
RuntimeError
RuntimeWarning
StandardError
StopIteration
SyntaxError
SyntaxWarning
SystemError
SystemExit
TabError
TypeError
UnboundLocalError
UnicodeDecodeError
UnicodeEncodeError
UnicodeError
UnicodeTranslateError
UnicodeWarning
UserWarning
ValueError
Warning
ZeroDivisionError
```

## 异常处理

### 什么是异常处理？

python解释器检测到错误，触发异常（也允许程序员自己触发异常）

程序员编写特定的代码，专门用来捕捉这个异常（这段代码与程序逻辑无关，与异常处理有关）

如果捕捉成功则进入另外一个处理分支，执行你为其定制的逻辑，使程序不会崩溃，这就是异常处理

### 为何要进行异常处理

python解析器去执行程序，检测到了一个错误时，触发异常，异常触发后且没被处理的情况下，程序就在当前异常处终止，后面的代码不会运行，谁会去用一个运行着突然就崩溃的软件。

所以你必须提供一种异常处理机制来增强你程序的健壮性与容错性 

### 如何进行异常处理

首先须知，异常是由程序的错误引起的，语法上的错误跟异常处理无关，必须在程序运行前就修正。

#### 通过if判断来进行异常处理

- if判断式的异常处理只能针对某一段代码，对于不同的代码段的相同类型的错误你需要写重复的if来进行处理。
- 在你的程序中频繁的写与程序本身无关，与异常处理有关的if，就像是在你的代码中到处拉屎。可读性极其的差
- 千万不要妄下定论if不能用来异常处理。如果你不服，那你想想在没有学习try...except之前，你的程序难道就没有异常处理，而任由其崩溃么

```
def test():
    print('test running')
choice_dic={
    '1':test
}
while True:
    choice=input('>>: ').strip()
    if not choice or choice not in choice_dic:continue #这便是一种异常处理机制啊
    choice_dic[choice]()
```

#### 使用try...except..来进行异常处理

##### 1、基本使用

```
语法：
try:
    被检测的代码块
except 异常类型：
    try中一旦检测到异常，就执行这个位置的逻辑
    
示例代码：
    
f=open('a.txt')
g=(line.strip() for line in f)
'''
next(g)会触发迭代f，依次next(g)就可以读取文件的一行行内容，无论文件a.txt有多大，同一时刻内存中只有一行内容。
提示：g是基于文件句柄f而存在的，因而只能在next(g)抛出异常StopIteration后才可以执行f.close()
'''

f=open('a.txt')

g=(line.strip() for line in f)
for line in g:
    print(line)
else:
    f.close()
    

try:
    f=open('a.txt')
    g=(line.strip() for line in f)
    print(next(g))
    print(next(g))
    print(next(g))
    print(next(g))
    print(next(g))
except StopIteration:
    f.close()
```

##### 2、异常类只能用来处理指定的异常情况，如果非指定异常则无法处理。

```
# 未捕获到异常，程序直接报错
 
s1 = 'hello'
try:
    int(s1)
except IndexError as e:
    print e
```

##### 3、针对多个异常类型我们可以写多个分支

```
s1 = 'hello'
try:
    int(s1)
except IndexError as e:
    print(e)
except KeyError as e:
    print(e)
except ValueError as e:
    print(e)
```

##### 4、万能异常(Exception)，可以捕获任意异常类型的异常。

```python
s1 = 'hello'
try:
    int(s1)
except Exception as e:
    print(e)
```

你可能会说既然有万能异常，那么我直接用上面的这种形式就好了，其他异常可以忽略

你说的没错，但是应该分两种情况去看

1.如果你想要的效果是，无论出现什么异常，我们统一丢弃，或者使用同一段代码逻辑去处理他们，那么骚年，大胆的去做吧，只有一个Exception就足够了。

```
s1 = 'hello'
try:
    int(s1)
except Exception,e:
    '丢弃或者执行其他逻辑'
    print(e)

#如果你统一用Exception，没错，是可以捕捉所有异常，但意味着你在处理所有异常时都使用同一个逻辑去处理（这里说的逻辑即当前expect下面跟的代码块）
```

2.如果你想要的效果是，对于不同的异常我们需要定制不同的处理逻辑，那就需要用到多分支了。当然也可以在多分支后来一个exception

```
s1 = 'hello'
try:
    int(s1)
except IndexError as e:
    print(e)
except KeyError as e:
    print(e)
except ValueError as e:
    print(e)
except Exception as e:
    print(e)
```

##### 5、异常的其他结构

```
s1 = 'hello'
try:
    int(s1)
except IndexError as e:
    print(e)
except KeyError as e:
    print(e)
except ValueError as e:
    print(e)
#except Exception as e:
#    print(e)
else:
    print('try内代码块没有异常则执行我')
finally:
    print('无论异常与否,都会执行该模块,通常是进行清理工作') 
```

else在try内代码执行正常的时候才会执行else里面的逻辑。如果try里面的代码块有异常的话那么就会执行对应的except的了。

那么finally其实就是不管你有没有异常都会执行finally下面的代码块，一般用于清理工作，那么什么是清理工作呢。比如客户端链接远端服务器，然后客户端崩了相当于tcp链接异常被关掉了，但是server端还为client端保存着这个已经无意义的tcp链接，这就是占用无用的资源，因此需要finally这种清理的操作来进行资源的释放。

##### 6、主动抛出异常，raise

```
#_*_coding:utf-8_*_
__author__ = 'Linhaifeng'

try:
    raise TypeError('类型错误')
except Exception as e:
    print(e)
```

##### 7、自定义异常

其实上面不管是TypeError还是什么属性不存在，他们的本质其实都是类，他们的父类是BaseException，如下：

```
print(TypeError.__bases__) # (<class 'Exception'>,)
```

因此我们也可以自己自定义一个错误异常类，继承BaseException就可以了。

```
#_*_coding:utf-8_*_

class EgonException(BaseException):
    def __init__(self,msg):
        self.msg=msg
    def __str__(self):
        return self.msg

try:
    raise EgonException('类型错误')
except EgonException as e:
    print(e) 
```

##### 8、断言

```
# assert 条件
 
assert 1 == 1
assert 1 == 2

等价于：
if not 1 == 1:
    raise AssertionError
```

在开发程序的时候，与其让它在运行的时候崩溃，不如在他出现错误条件的时候就崩溃返回错误这是断言的一个用法。

首先什么时候应该使用断言，没有特定的规则，断言应该用于：

- 防御型的编程
- 运行时检查程序逻辑
- 检查约定
- 程序常量
- 检查文档

针对断言的内容可以参考以下几篇文章：

[Python中不尽如人意的断言](http://www.cnblogs.com/cicaday/p/python-assert.html)

[Python使用断言的最佳时机](https://www.oschina.net/translate/when-to-use-assert)

[Python常用断言](http://www.jianshu.com/p/eea0b0e432da)

##### 9、try..except的方式比较if的方式的好处

try..except这种异常处理机制就是取代if那种方式，让你的程序在不牺牲可读性的前提下增强健壮性和容错性

异常处理中为每一个异常定制了异常类型（python中统一了类与类型，类型即类），对于同一种异常，一个except就可以捕捉到，可以同时处理多段代码的异常（无需‘写多个if判断式’）减少了代码，增强了可读性 

使用try..except的方式

1. 把错误处理和真正的工作分开来
2. 代码更易组织，更清晰，复杂的工作任务更容易实现；
3. 毫无疑问，更安全了，不至于由于一些小的疏忽而使程序意外崩溃了；

##### 10、if判断和try...except...的选择

- if本身就可以来处理异常，只不过if的方式，对于不同代码段的同一种异常，需要重复写多分支的if，而这段多分支if与真正的工作无关，写多了你的程序可读性就会及其的差。
- try..except的方式，只是python提供给你一种特定的语法结构去做这件事，对于不同代码的同一种异常，python为你定制了一中类型，一个expect就可以捕捉到

## 什么时候该使用异常处理

有的同学会这么想，学完了异常处理后，好强大，我要为我的每一段程序都加上try...except，干毛线去思考它会不会有逻辑错误啊，这样就很好啊，多省脑细胞===》2B青年欢乐多

try...except应该尽量少用，因为它本身就是你附加给你的程序的一种异常处理的逻辑，与你的主要的工作是没有关系的
这种东西加的多了，会导致你的代码可读性变差

而且异常处理本就不是你2b逻辑的擦屁股纸，只有在有些异常无法预知的情况下，才应该加上try...except，其他的逻辑错误应该尽量修正